codebase/src/services/completions.md

```diff
@@ -1,9 +1,62 @@
-## Completions in TS
+# Completions
 
-Completions come in through the language service, and most of the work starts in
-[`getCompletionEntriesFromSymbols`](https://github.com/Microsoft/TypeScript/blob/340f81035ff1d753e6a1f0fedc2323d169c86cc6/src/services/completions.ts#L305).
 
-### Completion Data
+Completions for TypeScript and JavaScript are provided by TypeScript's language service.
+When you are in the middle of typing something in an editor (or if you hit `Ctrl` + `Space` in VSCode), the editor sends a request to the TypeScript language service.
+Completions is responsible for answering that request with suggestions to *complete* what you are typing.
+
+
+## Overview
+
+Most of the implementation lives in the `src/services/completions.ts` file, and there are several steps the implementation goes through to answer a completion request.
+
+The entry point into completions is `getCompletionsAtPosition()`.
+As the name suggests, this function takes a `SourceFile` and a `position` as arguments (among other things), and it returns a [`CompletionInfo`](https://github.com/microsoft/TypeScript/blob/404a7d602df9c19d98d49e6a6bf2295e423be676/src/services/types.ts?#L1172-L1191) object with completions for that specific position.
+
+`CompletionInfo` has a few different properties, but we're mainly interested in the `entries: CompletionEntry[]` property, because a [`CompletionEntry`](https://github.com/microsoft/TypeScript/blob/404a7d602df9c19d98d49e6a6bf2295e423be676/src/services/types.ts?#L1220-L1249) encodes the suggestions returned.
+
+
+### Completion entry
+
+Some `CompletionEntry` properties and what they mean:
+
+* **name**: the name of that completion entry. Usually if the completion is for an identifier/keyword named `foo`, then the name of the entry is also going to be `foo`.
+* **insertText**: the text that is going to be inserted in the file, at the completion position, when the user accepts the suggestion corresponding to this completion entry.
+`insertText` is optional, and if it is not present, then `name` is the text that is inserted instead.
+* **isSnippet**: if this is true, then this completion entry is a snippet, and `insertText` is a snippet text.  
+e.g.:
+A completion snippet for declaring a method `foo`, with a tab stop (`${0}`) in its body:
+```ts
+{
+    isSnippet: true,
+    insertText: "foo() { ${0} }",
+}
+```
+becomes this in VSCode, when accepted (note the cursor position):
+![Screenshot of vscode with code `class Foo { foo() { | } }` in it.](../../screenshots/threeslash-refs.png)
+For more on snippets, see [Snippets in Visual Studio Code](https://code.visualstudio.com/docs/editor/userdefinedsnippets).
```

Comment:
Preserving the list hierarchy with a little Markdown magic:

````suggestion
* **isSnippet**: if this is true, then this completion entry is a snippet, and `insertText` is a snippet text.  
  e.g.:
  A completion snippet for declaring a method `foo`, with a tab stop (`${0}`) in its body:
  ```ts
  {
      isSnippet: true,
      insertText: "foo() { ${0} }",
  }
  ```
  becomes this in VSCode, when accepted (note the cursor position):
  ![Screenshot of vscode with code `class Foo { foo() { | } }` in it.](../../screenshots/threeslash-refs.png)
  For more on snippets, see [Snippets in Visual Studio Code](https://code.visualstudio.com/docs/editor/userdefinedsnippets).
````